README.md
# Brazil Monthly Deaths by City

This package exports a dataset with deaths number for all cities of Brazil since 2015. To obtain the most recent data, it also exports a web scraping function to get data every time you want.

[![Codecov Coverage](https://img.shields.io/codecov/c/github/pedrobern/brazil-monthly-deaths-by-city/master.svg?style=flat-square)](https://codecov.io/gh/pedrobern/brazil-monthly-deaths-by-city/)
[![Build Status](https://travis-ci.com/pedrobern/brazil-monthly-deaths-by-city.svg?branch=master)](https://travis-ci.com/pedrobern/brazil-monthly-deaths-by-city)
[![Pypi](https://img.shields.io/pypi/v/brazil-monthly-deaths.svg)](https://pypi.org/project/brazil-monthly-deaths/)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](https://github.com/pedrobern/brazil-monthly-deaths-by-city/blob/master/CONTRIBUTING.md)


## Table of Contents

- [Installation](#Installation)
- [Usage](#Usage)
- [Data example](#Data-example)
- [API](#API)
  - [Dataframes](#Dataframes)
  - [brazil_deaths](#brazil-deaths)
  - [update_df](#update-df)
  - [get_city_id](#get-city-id)

<div id="Installation"></div>

## Installation

### To work only with the provided datasets

```bash
pip install brazil-monthly-deaths
```

### To get the most recent data

If you want to **get the most recent data**, probably not available on this package, since the deaths numbers change every day, install the chrome driver in order to use selenium, you can see more information in the [selenium documentation](https://selenium-python.readthedocs.io/installation.html#drivers) and the [chrome driver download page](https://sites.google.com/a/chromium.org/chromedriver/downloads).

<div id="Usage"></div>

## Usage

> :warning: When getting new data, this package will access an official website for new records, but as it is made with web scraping, if the html is updated, this package will no longer work, needing to be updated accordingly. Please if you experience any kind of failure, open an issue.

> Assuming you have installed the chrome driver

```python
from brazil_monthly_deaths import brazil_deaths, data, update_df

# data is the data from 2015 to 2020
print(data)

# Everyday there are new records,
# so you should get the most recent data.
# Depending on your internet connection
# it may take up to 6 minutes for each month
# if you run for all states. Consider selecting
# only the states you want to work on.
new_data = brazil_deaths(years=[2020], months=[5])

# update the lagging data provided by this package
current_data = update_df(data, new_data)
print(current_data)
```

See the [jupyter example](https://github.com/PedroBern/brazil-monthly-deaths-by-city/tree/master/jupyter_example) to generate the following plot:

![plot example](./jupyter_example/plot.png)

The example running:

![demo example](./jupyter_example/demo.gif)
<div id="Data-example"></div>

## Data example

| city_id  | year | month | region    | state          | city        | deaths |
|----------|------|-------|-----------|----------------|-------------|--------|
| 3516805  | 2020 | 1     | Southeast | Rio de Janeiro | Tracunhaém  | 8      |
| 21835289 | 2020 | 1     | Southeast | Rio de Janeiro | Trindade    | 13     |
| 10791950 | 2020 | 1     | Southeast | Rio de Janeiro | Triunfo     | 16     |
| 81875827 | 2020 | 1     | Southeast | Rio de Janeiro | Tupanatinga | 18     |
| 99521011 | 2020 | 1     | Southeast | Rio de Janeiro | Tuparetama  | 4      |


---

<div id="API"></div>

## API

<div id="Dataframes"></div>

### Dataframes

This package exports some [pandas](https://github.com/pandas-dev/pandas) dataframe with the following columns:

- city_id : unique integer from state and city,
- year : from 2015 to 2020,
- month : from 1 to 12,
- region : [North, Northeast, South, Southeast, Center_West],
- state : one of the 27 states of Brazil, including country capital,
- city : city name
- deaths : number os deaths

```python
from brazil_monthly_deaths import (
  data, # full data
  data_2015,
  data_2016,
  data_2017,
  data_2018,
  data_2019,
  data_2020 # always out of date, you need to update it
)
```

<div id="brazil-deaths"></div>

### brazil_deaths

You can use this function to scrap new data directly from the [Civil Registry Offices website](https://transparencia.registrocivil.org.br/registros). Just make sure you have installed the chrome driver, as pointed above.

Official note about the legal deadlines:

> The family has up to 24 hours after the death to register the death in the Registry, which, in turn, has up to five days to perform the death registration, and then up to eight days to send the act done to the National Information Center of the Civil Registry ( CRC Nacional), which updates this platform.

It means: **The last 13 days are always changing.**

```python
from brazil_monthly_deaths import brazil_deaths
```

Since it will access an external website, it will depend on your internet connection and world location. Consider selecting only the `states` you want to work on. For each month, for all states it may take up to 6 min to run for a single year.

```python
df = brazil_deaths(
    years=[2015, 2016, 2017, 2018, 2019, 2020],
    months=range(1, 13, 1),
    regions=_regions_names,
    states=_states,
    filename="data",
    return_df=True,
    save_csv=True,
    verbose=True,
    *args,
    **kwargs
)  
```

The `_regions_names` is:

```python
["North", "Northeast", "South", "Southeast", "Center_West"]
```

The `_states` is:

```python
[
  "Acre", "Amazonas", "Amapá", "Pará", 
  "Rondônia", "Roraima", "Tocantins", "Paraná",
  "Rio Grande do Sul", "Santa Catarina", "Espírito Santo",
  "Minas Gerais", "Rio de Janeiro", "São Paulo",
  "Distrito Federal", "Goiás", "Mato Grosso do Sul",
  "Mato Grosso", "Alagoas", "Bahia", "Ceará",
  "Maranhão", "Paraíba", "Pernambuco",
  "Piauí", "Rio Grande do Norte", "Sergipe"
]
```

The `*args` and `**kwargs` are passed down to `df.to_csv(..., *args, **kwargs)`

<div id="update-df"></div>

### update_df

Use this function after you have scraped recent data from the Civil Registry Offices website to update the data provided in this package.

```python
from brazil_monthly_deaths import brazil_deaths, data, update_df

new_data = brazil_deaths(years=[2020], months=[5])
current_data = update_df(data, new_data)
```

It basically put the new data below the old data in the dataframe, then remove the duplicates (excluding deaths) keeping the most recent entries.

<div id="get-city-id"></div>

### get_city_id

Get the unique id of the combination of the `state` and `city`.

```python
from brazil_monthly_deaths import get_city_id

sao_paulo_id = get_city_id(state='São Paulo', city='São Paulo')

print(sao_paulo_id) # 89903871
```


CHANGES.md
# Changelog

## v1

### v1.4.0

- Update web scrap accordingly to website html updates.

### v1.3.0

- Add `get_city_id` function to module exports.

### v1.2.0

- Update 2020 data.

### v1.1.0

- Check if both dataframes have same columns and dtypes before updating in `update_df`.

### v1.0.1

- Update README on PyPi.

### v1.0.0

- First release.
- Include data up to 05/20/20


tests/test_web_scraping.py
import pytest

from brazil_monthly_deaths import brazil_deaths
from brazil_monthly_deaths.exceptions import WrongInput


def test_raise_invalid_year():
    pytest.raises(WrongInput, brazil_deaths, years=[2000])


def test_raise_invalid_month():
    pytest.raises(WrongInput, brazil_deaths, months=[13])


def test_raise_invalid_state():
    pytest.raises(WrongInput, brazil_deaths, states=["invalid"])


def test_raise_invalid_region():
    pytest.raises(WrongInput, brazil_deaths, regions=["invalid"])


tests/test_update_new_data.py
from unittest import TestCase
import pandas as pd
import numpy as np

from brazil_monthly_deaths.exceptions import WrongInput
from brazil_monthly_deaths import (
    update_df,
    data_2015,
)

columns = ["city_id", "year", "month", "region", "state", "city", "deaths"]


class TestUpdateNewData(TestCase):
    def setUp(self):
        fake_data = [[82280746, 2015, 1, "North", "Acre", "Brasiléia", 10]]
        self.fake_df = pd.DataFrame(np.asarray(fake_data), columns=columns)

    def test_update_df_wrong_dtype(self):
        with self.assertRaises(WrongInput):
            updated_df = update_df(data_2015, self.fake_df)

    def test_update_df(self):
        self.fake_df["city_id"] = pd.to_numeric(self.fake_df["city_id"])
        self.fake_df["year"] = pd.to_numeric(self.fake_df["year"])
        self.fake_df["month"] = pd.to_numeric(self.fake_df["month"])
        self.fake_df["deaths"] = pd.to_numeric(self.fake_df["deaths"])
        updated_df = update_df(data_2015, self.fake_df)
        assert len(data_2015) == len(updated_df)


brazil_monthly_deaths/web_scraping.py
"""
Web Scraping of Brazil Deaths
"""
import hashlib
from selenium import webdriver
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.common.by import By
from selenium.webdriver import ActionChains
import numpy as np
import pandas as pd

from .exceptions import WrongInput

# From where the data comes
url = "https://transparencia.registrocivil.org.br/registros"

# State names of Brazil
north_states = ["Acre", "Amazonas", "Amapá", "Pará", "Rondônia", "Roraima", "Tocantins"]
south_states = ["Paraná", "Rio Grande do Sul", "Santa Catarina"]
southeast_states = ["Espírito Santo", "Minas Gerais", "Rio de Janeiro", "São Paulo"]
center_west_states = ["Distrito Federal", "Goiás", "Mato Grosso do Sul", "Mato Grosso"]
northeast_states = [
    "Alagoas",
    "Bahia",
    "Ceará",
    "Maranhão",
    "Paraíba",
    "Pernambuco",
    "Piauí",
    "Rio Grande do Norte",
    "Sergipe",
]

# Regions of Brazil (defined above)
_regions = {
    "North": north_states,
    "Northeast": northeast_states,
    "South": south_states,
    "Southeast": southeast_states,
    "Center_West": center_west_states,
}

# Region names
_regions_names = ["North", "Northeast", "South", "Southeast", "Center_West"]

# dataframe
df_states = pd.DataFrame(dict([(k, pd.Series(v)) for k, v in _regions.items()]))
df_states = pd.melt(
    df_states, value_vars=_regions_names, var_name="region", value_name="state"
)
df_states = df_states.dropna()
df_states = df_states.reset_index(drop=True)

# flatten states
_states = [j for i in [i for i in _regions.values()] for j in i]

# Months in portuguese
_months = [
    "Janeiro",
    "Fevereiro",
    "Março",
    "Abril",
    "Maio",
    "Junho",
    "Julho",
    "Agosto",
    "Setembro",
    "Outubro",
    "Novembro",
    "Dezembro",
]

# Years with data available
_years = [2020, 2019, 2018, 2017, 2016, 2015]

# paths on the webpage
# if the webpage is updtaed and change these,
# the scrip will no longer work and need to be
# updated accordingly
year_path = '//fieldset[@id="datePickrGroup"]/div'
month_path = '//legend[text()="Mês"]/ancestor::fieldset/div'
state_path = '//label[text()="Estado"]/parent::div/div'
search_path = '//button[normalize-space()="Pesquisar"]'
city_name_row_path = "//tbody/tr/td[@aria-colindex=1]"
city_deaths_row_path = "//tbody/tr/td[@aria-colindex=2]"
table_id = "__BVID__170"
next_page_path = '//li/a/span[text()="›"]/ancestor::li'
empty_table_path = '//tr[@class="b-table-empty-row"]'
deaths_input_path = '//input[@value="death"]/parent::div'

# pandas df headers
header = ["city_id", "year", "month", "region", "state", "city", "deaths"]


def open_page():
    """
    open the webpage
    """
    browser.get(url)
    assert "Portal da Transparência" in browser.title


def select_deaths():
    """
    select the deaths radio input
    """
    browser.find_element_by_xpath(deaths_input_path).click()


def select_field(path, data):
    """
    select the target data in the form
    """
    browser.find_element_by_xpath(path).click()
    browser.find_element_by_xpath(
        '//span[text()="' + str(data) + '"]/ancestor::li'
    ).click()


def press_search():
    """
    press the search button
    """
    wait.until(EC.element_to_be_clickable((By.XPATH, search_path)))
    button = browser.find_element_by_xpath(search_path)
    actions = ActionChains(browser)
    actions.move_to_element(button).click(button).perform()


def select_fields(year, month, state):
    """
    select the year, month and state, then press the search
    """
    month = _months[month - 1]  # convert month from number to portuguese word
    select_field(year_path, year)
    select_field(month_path, month)
    select_field(state_path, state)
    press_search()


def get_table_data():
    """
    return city names and deaths number
    if the table is empty, return False
    """
    wait.until(
        EC.presence_of_element_located(
            (By.XPATH, empty_table_path + "|" + city_name_row_path)
        )
        # or EC.presence_of_element_located((By.XPATH, city_name_row_path))
    )
    empty_table = len(browser.find_elements_by_xpath(empty_table_path)) != 0

    if not empty_table:
        cities = [
            item.text for item in browser.find_elements_by_xpath(city_name_row_path)
        ]
        deaths = [
            item.text for item in browser.find_elements_by_xpath(city_deaths_row_path)
        ]
        return {"cities": cities, "deaths": deaths}
    else:
        return False


def next_page():
    """
    if there is the next page button, press it
    """
    try:
        browser.find_element_by_xpath(next_page_path).click()
        return True
    except:
        return False


def save_data(data, year, filename="data", *args, **kwargs):
    """
    save the data as csv
    """
    filename += "_" + year + ".csv"
    data.to_csv(filename, index=False, header=header, *args, **kwargs)


def get_city_id(state, city):
    s = state + " " + city
    hash = int(hashlib.sha256(s.encode("utf-8")).hexdigest(), 16) % 10 ** 8
    return hash


def brazil_deaths(
    years=_years,
    months=range(1, 13, 1),
    regions=_regions_names,
    states=_states,
    filename="data",
    return_df=True,
    save_csv=True,
    verbose=True,
    *args,
    **kwargs
):
    """
        save the tidy data as csv for each year and/or return a pandas df
    """

    # check the inputs
    if not set(years).issubset(set(_years)):
        raise WrongInput("Year must be between 2015 and 2020")
    if not set(months).issubset(set(range(1, 13, 1))):
        raise WrongInput("Months must be between 1 and 12")
    if not set(states).issubset(set(_states)):
        raise WrongInput("Invalid states")
    if not set(regions).issubset(set(_regions_names)):
        raise WrongInput("Invalid regions")

    # order the months
    months = sorted(months)

    # filter rows based on input and group by regions
    _df_states = df_states.query("region in @regions")
    _df_states = _df_states.query("state in @states")
    _df_states = _df_states.groupby("region")

    # do the web scraping
    global browser, wait
    browser = webdriver.Chrome()
    wait = WebDriverWait(browser, 60)

    open_page()
    select_deaths()
    if return_df:
        combined_df = pd.DataFrame(columns=header)
    for year in years:
        if verbose:
            print("year:", year)
        year_data = []
        for month in months:
            if verbose:
                print("month:", month)
            for region, df_group in _df_states:
                if verbose:
                    print("region:", region)
                for index, row in df_group.iterrows():
                    state = row["state"]
                    if verbose:
                        print("state:", state)
                    select_fields(year, month, state)
                    next_p = True
                    page = 1
                    state_cities = []
                    state_deaths = []
                    while next_p:
                        if verbose:
                            print("page:", page)
                        d = get_table_data()
                        if d is not False:
                            state_cities += d["cities"]
                            state_deaths += d["deaths"]
                        next_p = next_page()
                        page += 1
                    for c, city in enumerate(state_cities):
                        city_id = get_city_id(state, city)
                        year_data.append(
                            [
                                city_id,
                                year,
                                month,
                                region,
                                state,
                                city,
                                state_deaths[c],
                            ]
                        )
        if len(year_data) > 0:
            df = pd.DataFrame(np.asarray(year_data), columns=header)
            if save_csv:
                save_data(df, str(year), filename=filename, *args, **kwargs)
            if return_df:
                combined_df = combined_df.append(df)
    browser.quit()
    if return_df:
        if len(combined_df) > 0:
            combined_df["city_id"] = pd.to_numeric(combined_df["city_id"])
            combined_df["year"] = pd.to_numeric(combined_df["year"])
            combined_df["month"] = pd.to_numeric(combined_df["month"])
            combined_df["deaths"] = pd.to_numeric(combined_df["deaths"])
            return combined_df
        else:
            return None
